src/lib.rs

```diff
@@ -694,7 +837,26 @@ impl<T> ThinVec<T> {
             (*ptr).set_cap(new_cap);
             self.ptr = NonNull::new_unchecked(ptr);
         } else {
-            self.ptr = header_with_capacity::<T>(new_cap);
+            let mut new_header = header_with_capacity::<T>(new_cap);
+
+            // If we get here and have a non-zero len, then we must be handling
+            // a gecko auto array, and we have items in a stack buffer. We shouldn't
+            // free it, but we should memcopy the contents out of it and mark it as empty.
+            //
+            // This leans on the assumption that T has a trivial move constructor,
+            // which nsTArray already requires. Furthermore, we assume C++ won't try
```

Comment:
This sentence does not seem entirely correct to me.  Are you trying to claim that `T` has a trivial move constructor, which we require in the case of trying to make `nsTArray` and `ThinVec` interoperate?  That claim makes more sense to me.

Reply:
For most types, `nsTArray<T>` already assumes that `T` has a trivial move constructor, so that it can use `memcpy` to move items around in the array. All types in Rust fit that requirement.

There are some types in C++ which are opted-out of that requirement, using `nsTArray_RelocateUsingMoveConstructor` as the `nsTArray_ReloactionStrategy`. Those types are fundamentally incompatible with Rust.

Reply:
It's certainly required for interop since Rust can't see the move ctors. That said I thought it was also the case that nsTArray required trivial moves, in the sense that you could e.g. realloc the nsTArray buffer?

Reply:
Like @mystor said, you can specify what you want `nsTArray` to do and the right thing happens when you resize, `realloc` or `malloc` as appropriate: https://searchfox.org/mozilla-central/rev/26b13464c2beb26e0d864d561c30e817a85c348a/xpcom/ds/nsTArray-inl.h#195-214

Reply:
(There's a good chance "trivial" is the wrong word here, as it's been a while since I've had to discuss ctor properties)

Reply:
I specifically mean the situation where move+delete are implemented such that (pseudo cpp):

```
auto dest = move(src)
delete(src)
```

can be rewritten to 

```
memcopy(&dest, src)
forget(src)
```

That is, when you move from src to dest: 
* dest becomes a bitwise copy of src (almost always true)
* src becomes a value whose dtor is a no-op (almost always true)

Reply:
The ~only types which don't satisfy this property have significant addresses. e.g. they're intrusive (linked list on stack) or registered with a supervisor (swift/objc weak references).

Reply:
> For most types, nsTArray<T> already assumes that T has a trivial move constructor, so that it can use memcpy to move items around in the array. 

That's not correct:
* There are instantiations of `nsTArray<T>` for `T` which has no copy or move constructor at all. 
* The requirement is not related to moving at all, but to relocating, which is not the same thing. (On the language level, only C++20 will introduce the concept of (trivial) relocatability, IIRC). This is of practical relevance: E.g. `UniquePtr<T>` has a non-trivial move constructor (if it just memcopied, we would end up with two UniquePtr instances owning the same memory, which would lead to a double free when both instances are destroyed), but it is trivially relocatable (the source of the relocation will simply be discarded, and its destructor will never be called).

Reply:
Author of [P1144](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p1144r10.html) here, just popping in three years later to say that (1) yes, you have the notion of "trivially relocatable" exactly right, (2) it didn't make C++20 and isn't likely even to make C++26 without a lot of help from the public, and (3) there's a PR out _this week_ to add P1144-compatible `__is_trivially_relocatable(T)` support to the Clang C++ compiler. If you're interested in the area, I encourage you to [take a look at that PR](https://github.com/llvm/llvm-project/pull/84621) and leave a comment in support. Thank you! :)